rfcs/proposed/numa_support/increase-numa-support-availability.org

```diff
@@ -0,0 +1,115 @@
+by the library as part of its initialization stage. In turn, each ~tbbbind~ has
+a hard dependency, i.e., relies on load-time linking, on a concrete version of
+the HWLOC library [1, 2]. The soft dependency of oneTBB on ~tbbbind~ allows the
+library to continue its execution even if the system loader is unable to resolve
+the hard dependency on HWLOC for ~tbbbind~. In this case, the HW topology is not
+discovered and the machine is seen as if all CPU cores were uniform, which is
+the default TBB behavior when NUMA constraints are not used. Thus, the following
+code returns meaningless values as these values are just ignored by oneTBB:
+
+#+begin_src C++
+std::vector<oneapi::tbb::numa_node_id> numa_nodes = oneapi::tbb::info::numa_nodes();
+std::vector<oneapi::tbb::core_type_id> core_types = oneapi::tbb::info::core_types();
+#+end_src
+
+An error is also not reported and the client code that uses NUMA support
+facilities may continue running expecting it to work as it was intended. Such
+behavior is not readily noticable by developers that use oneTBB and this
+represents the main problem with the current behavior.
+
+Having a dependency on a shared HWLOC library has a number of advantages:
+1. Code reuse with all of the positive consequences out of this. That's the
+   primary purpose of shared libraries.
+2. Sharing HWLOC context between its clients. This avoids performing the same
+   operations repeatedly with identical results.
+3. A drop-in replacement. Users are able to use their own version of HWLOC
+   without recompilation of oneTBB.
+
+The only disadvantage from depending on HWLOC library dynamically is that the
+developers that use oneTBB's NUMA support API need to make sure the library is
+available and can be found by oneTBB. Depending on the distribution model of a
+developer's code, this is achieved either by:
+1. Asking the end user to have necessary version of a dependency pre-installed.
+2. Bundling necessary HWLOC version together with other pieces of a product
+   release.
+
+However, the requirement to fulfill one of the above steps for the NUMA API to
+start paying off may be considered as an incovenience and, what is more
+important, it is not always obvious that one of these steps is needed.
+Especially, due to silent behavior in case HWLOC library cannot be found in the
+environment.
+
+This proposal suggests an improvement to reduce the effect of the disadvantage
+being dependent on a dynamic version of HWLOC library by having it linked
+statically with one of the ~tbbbind~ libraries that are distributed together
+with oneTBB, yet leaving possibility to specify another version of HWLOC library
+if users see the need.
+
+[1] [[https://www.open-mpi.org/projects/hwloc/][HWLOC project main page]]
+
+[2] [[https://github.com/open-mpi/hwloc][HWLOC project repository on GitHub]]
+
+* Proposal
+1. Introduce new variant of the ~tbbbind~ library with the name ~tbbbind_static~
```

Comment:
Possibly an open question: Could it make sense to drop support of HWLOC 1.x and use the name `tbbbind` (with no suffixes) for this new statically linked variant? That would also avoid the need to ship one more binary.

Reply:
I rather disagree about it being unrelated, as it directly addresses one of the shortcomings of the proposal.